### ISSUE TITLE
Add Humanization

### ISSUE THREAD

**Issue by JeroenDeDauw** [ProfessionalWiki/EDTF#18]
Formats:

* [x] `1975-07-01`: July 1, 1975
* [x] `1975-12`: December 1975
* [x] `1975`: 1975
* [x] `0043`:      43
* [x] `1975-22`: Summer 1975.
* [x] `XXXX-12-XX`: December
* [x] `XXXX-XX-12`: 12th
* [x] `XXXX-12-11`: December 11th
* [x] `2020-XX-11`: 11th of unknown month, 2020
* [x] `2019?`:    ca. 2019
* [x] `-1234`:    1234 av.J.-C.
* [x] `1981/1985`: 1981 to 1985
* [x] `1981/..`: 1981 or later
* [x] `../1985`: 1985 or earlier
* [x] `1981/`: 1981 to unknown
* [x] `197x`: 1970s (blocked by https://github.com/ProfessionalWiki/EDTF/issues/21)
* [x] `19xx`: 1900s or 19 siècle (blocked by https://github.com/ProfessionalWiki/EDTF/issues/21)
* [x] Time of day (23:20:30 UTC-11:45 April 12th, 1985)

Languages:

* [x] English
* [ ] French

Integrations:

* [x] Wikibase
* [ ] [TranslateWiki](https://github.com/ProfessionalWiki/EDTF/issues/20)

**Comment by JeroenDeDauw** [comment]
@mzeinstra I'm wondering if there are any

**Comment by JeroenDeDauw** [comment]
@mzeinstra currently I am implementing the examples from the PDF plus what I think makes sense. If there is more of a specification, now is the time to share it. Else it is me guessing at which formats should be humanized and what the wording in the humanization should be. (Which might of course be good enough.)

**Comment by JeroenDeDauw** [comment]
@mzeinstra can you also confirm this should initially work with both English and French? Given the complexity of the translation logic, we unfortunately cannot use TranslateWiki.

**Comment by mzeinstra** [comment]
I see that you are making a class per language, could you not use i18n structure that provides the language as a string upon the creation of a generic class (with fallback to 'en'). Than we could use TranslateWiki to generate the PO files right?

**Comment by JeroenDeDauw** [comment]
Many strings could be moved to TranslateWiki, but not all "language specific handling". The logic between languages can differ, and TranslateWiki can't handle that. You can just give it a string with some parameters. It can't do complicated logic on those parameters, and having a dozen parameters or playing lego is discouraged.

Total work to have English and French with the current approach is definitely less than integrating with TranslateWiki. But if support for many languages is important, then we can of course work on that.

**Comment by JeroenDeDauw** [comment]
Maybe I was getting ahead of myself with edge case support. I'll investigate TranslateWiki integration cost implications.

Still would be good to get an idea of how many languages are needed for this project.

Another question: how should fallback work? Say the language is French, and there is no French message. My suspicion is that we then want to fall back to showing the EDTF (which also happens when there is no humanization for the EDTF format), rather than to English. Is that correct?

**Comment by JeroenDeDauw** [comment]
![image](https://user-images.githubusercontent.com/146040/107423989-6228d980-6b1d-11eb-9ad2-7d5a8e01b9ec.png)

![image](https://user-images.githubusercontent.com/146040/107424008-68b75100-6b1d-11eb-9964-75c3cb16578f.png)

**Comment by mzeinstra** [comment]
I'll chase the network on the number of languages. Luxembourg's working language is French, some older records might be in German as well. Luxembourgish standardisation is not used in formal communication if I recall correctly, but I expect this is a nice to have. English is a nice to have, but having that might make development easier.

I don't want to overstep, but given the modular representation of EDTF in your library would it not be possible to have a strftime kind of configuration sentence(s)? That way we don't have to develop an entire new class per language.

**Comment by mzeinstra** [comment]
I just saw your screenshot. I like it.

In our call we said that we would also display the raw EDTF string next to the humanised text.

**Comment by JeroenDeDauw** [comment]
> I don't want to overstep, but given the modular representation of EDTF in your library would it not be possible to have a strftime kind of configuration sentence(s)? That way we don't have to develop an entire new class per language.

You are right that needing to have one of these classes per language is not ideal. There is a lot of logic in there and right now it looks like most or all would be the same for the languages we care about. My current plan is to switch to a message-key based system. That would leave us with 1 class and a translation file per language. I suspect that is the same, or similar to, what you have in mind with strftime.

**Comment by JeroenDeDauw** [comment]
> In our call we said that we would also display the raw EDTF string next to the humanised text.

Thanks. I forgot about that and was thinking yesterday it would be nice to have this but did not get round to asking you about it yet. Tracked now at https://github.com/ProfessionalWiki/WikibaseEdtf/issues/2

**Comment by JeroenDeDauw** [comment]
@mzeinstra we have full level 0 and level 1 support now. ([tests that give an idea of supported formats](https://github.com/ProfessionalWiki/EDTF/blob/master/tests/Unit/Humanize/StringHumanizerTest.php#L20)) These level 2 cases are not supported yet, do we need them, and what is their priority?

* Sets
* Significant digits (`1950S2`: some year between 1900 and 1999, estimated to be 1950)
* Group Qualification (`2004-06~-11`: year and month approximate)
* Qualification of Individual Component (`?2004-06-~11`: year uncertain; month known; day approximate)

**Comment by JeroenDeDauw** [comment]
The answer: sets are important

Set humanization tracked by https://github.com/ProfessionalWiki/EDTF/pull/32
